src/helpers/privacy_budget.rs

```diff
@@ -100,43 +86,46 @@ impl PrivacyBudget {
     /// Budget is increased for each key atomically in Redis
     fn increase_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<bool> {
         let key = packet.to_string();
-        let (budget_consumed,): (u32,) =
-            transaction(&mut self.redis_connection, &[key.clone()], |con, pipe| {
-                let val: u32 = con.get(&key).unwrap_or(0);
-                // increment
-                pipe.set(&key, val + budget).ignore().get(&key).query(con)
-            })?;
-        if budget_consumed > self.get_total_privacy_budget() {
-            Ok(false)
-        } else {
-            Ok(true)
-        }
-    }
+        let total_privacy_budget = self.get_total_privacy_budget();
 
-    /// Budget is decreased for each ineligible key in Redis
-    fn decrease_budget_consumed(&mut self, packet: &BlindedMatchKey, budget: u32) -> Res<()> {
-        self.redis_connection.decr(packet.to_string(), budget)?;
-        Ok(())
-    }
-}
-
-impl From<RedisError> for Error {
-    fn from(_e: RedisError) -> Self {
-        Error::RedisError
+        Ok(transaction(
+            &mut self.redis_connection,
+            &[key.clone()],
+            |con, pipe| {
+                let val: u32 = con.get(&key).unwrap_or(0);
+                if val + budget > total_privacy_budget {
+                    pipe.set(&key, val).ignore().query::<()>(con)?;
+                    Ok(Some(false))
+                } else {
+                    // increment
+                    pipe.set(&key, val + budget)
+                        .ignore()
+                        .query::<()>(con)?;
```

Comment:
So this still fails if there are concurrent executions, so it might be better to use incr() rather than set() here.  That won't stop a race from happening that results in allowing a request that should be blocked, but it will at least ensure that we don't lose the budget that has been spent.

Note that the transaction handling in the redis crate is extraordinarily loose if you are used to ACID.

I don't know how to stop this from happening except to have this fail closed instead with incr, get, check, but that requires a rollback if the increment blows the budget.

Reply:
1. Redis doesnt support rollbacks so that is out. 
2 .A transaction would watch over the key for any change and will abort in case someone else changes the value. In that case, we would receive a "false" and discard that packet from being considered while also not burning the budget. --> this should take care of concurrency issues
3. However, if some error happens while executing the transaction, the budget might be burned while we still discard the packet --> Is the concern on this?

Reply:
Based on https://redis.io/docs/manual/transactions/#watchcommandswatch-explained looks like WATCH would not modify the budget consumed for a key in case of any concurrency issues. Due to this, going to keep this implementation